### PR TITLE
FOUNDATIONS: Add ShoppingItem

### PR DESCRIPTION
**Pull request by tharaniannamalai**

### CODE REVIEW

File: ShoppingList.ConsoleApp/Models/ShoppingItems/Exceptions/FailedShoppingItemServiceException.cs [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+using System.Text;
+using System.Threading.Tasks;
+using Xeptions;
+
+namespace ShoppingList.ConsoleApp.Models.ShoppingItems.Exceptions
+{
+    public class FailedShoppingItemServiceException : Xeption
+    {
+        public FailedShoppingItemServiceException(Exception innerException)
+            : base(message: "Failed shopping item service error occured, contact support.",
```
Comment (ShriHumrudha) [inline ShoppingList.ConsoleApp/Models/ShoppingItems/Exceptions/FailedShoppingItemServiceException.cs]: occurred - double r, check-in others too

File: ShoppingList.ConsoleApp/Models/ShoppingItems/Exceptions/InvalidShoppingItemException.cs [hunk 2]
```diff
@@ -0,0 +1,15 @@
...
+// ------------------------------------------------
+
+using Xeptions;
+
+namespace ShoppingList.ConsoleApp.Models.ShoppingItems.Exceptions
+{
+    public class InvalidShoppingItemException : Xeption
+    {
+        public InvalidShoppingItemException()
+            : base(message: "Shopping Item is invalid, fix the errors and try again.")
```
Comment (ShriHumrudha) [inline ShoppingList.ConsoleApp/Models/ShoppingItems/Exceptions/InvalidShoppingItemException.cs]: small i in item

File: ShoppingList.ConsoleApp/Models/ShoppingItems/Exceptions/ShoppingItemServiceException.cs [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+using System.Text;
+using System.Threading.Tasks;
+using Xeptions;
+
+namespace ShoppingList.ConsoleApp.Models.ShoppingItems.Exceptions
+{
+    public class ShoppingItemServiceException : Xeption
+    {
+        public ShoppingItemServiceException(Xeption innerException)
+            : base(message: "ShoppingItem service error occured, contact support.",
```
Comment (ShriHumrudha) [inline ShoppingList.ConsoleApp/Models/ShoppingItems/Exceptions/ShoppingItemServiceException.cs]: Shopping item

File: ShoppingList.ConsoleApp/Services/Foundations/ShoppingItems/ShoppingItemService.Exception.cs [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+        }
+
+        private ShoppingItemValidationException CreateAndLogValidationException(Xeption exception)
+        {
+            var shoppingItemValidationException = new ShoppingItemValidationException(exception);
+            this.loggingBroker.LogError(shoppingItemValidationException);
+
+            return shoppingItemValidationException;
+        }
+        private ShoppingItemServiceException CreateAndLogServiceException(Xeption exception)
```
Comment (ShriHumrudha) [inline ShoppingList.ConsoleApp/Services/Foundations/ShoppingItems/ShoppingItemService.Exception.cs:47]: New line space required, verify in other places too

File: ShoppingList.ConsoleApp/Services/Foundations/ShoppingItems/ShoppingItemService.Exception.cs [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+        {
+            var shoppingItemServiceException = new ShoppingItemServiceException(exception);
+            this.loggingBroker.LogError(shoppingItemServiceException);
+
+            return shoppingItemServiceException;
+        }
+
```
Comment (ShriHumrudha) [inline ShoppingList.ConsoleApp/Services/Foundations/ShoppingItems/ShoppingItemService.Exception.cs]: Remove extra line

File: ShoppingList.ConsoleApp/Services/Foundations/ShoppingItems/ShoppingItemService.cs [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+        private readonly IStorageBroker storageBroker;
+        private readonly ILoggingBroker loggingBroker;
+        public ShoppingItemService(
+            IStorageBroker storageBroker,
+            ILoggingBroker loggingBroker)
+        {
+            this.storageBroker = storageBroker;
+            this.loggingBroker = loggingBroker;
+        }
+        /*public ShoppingItem AddShoppingItem(ShoppingItem shoppingItem)
```
Comment (ShriHumrudha) [inline ShoppingList.ConsoleApp/Services/Foundations/ShoppingItems/ShoppingItemService.cs]: Remove the commented code

File: ShoppingList.Tests.Unit/Services/Foundations/ShoppingItems/ShoppingItemServiceTests.Exceptions.Add.cs [hunk 7]
```diff
@@ -0,0 +1,56 @@
...
+
+            var failedShoppingItemServiceException = 
+                new FailedShoppingItemServiceException(serviceException);
+
+            var expectedShoppingItemServiceException =
+                new ShoppingItemServiceException(failedShoppingItemServiceException);
+
+            this.storageBrokerMock.Setup(broker =>
+                broker.InsertShoppingItem(It.IsAny<ShoppingItem>()))
+                .Throws(serviceException);
```
Comment (ShriHumrudha) [inline ShoppingList.Tests.Unit/Services/Foundations/ShoppingItems/ShoppingItemServiceTests.Exceptions.Add.cs]: Need one more tab here, check-in other places too

File: ShoppingList.Tests.Unit/Services/Foundations/ShoppingItems/ShoppingItemServiceTests.Logic.Add.cs [hunk 8]
```diff
@@ -0,0 +1,44 @@
...
+            ShoppingItem actualShoppingItem = this.shoppingItemService.AddShoppingItem(inputShoppingItem);
+
+            // then
+            actualShoppingItem.Should().BeEquivalentTo(expectedShoppingItem);
+
+            this.storageBrokerMock.Verify(broker =>
+                broker.InsertShoppingItem(inputShoppingItem),
+                    Times.Once);
+
+            this.loggingBrokerMock.VerifyNoOtherCalls();
```
Comment (ShriHumrudha) [inline ShoppingList.Tests.Unit/Services/Foundations/ShoppingItems/ShoppingItemServiceTests.Logic.Add.cs]: Check the ordering here

File: ShoppingList.Tests.Unit/Services/Foundations/ShoppingItems/ShoppingItemServiceTests.Validations.Add.cs [hunk 9]
```diff
@@ -0,0 +1,90 @@
...
+
+        [Theory]
+        [InlineData(null)]
+        [InlineData("")]
+        [InlineData("  ")]
+        public void ShouldThrowValidationExceptionOnAddIfShoppingItemIsInvalidAndLogIt(
+            string invalidText)
+        {
+            // given
+            ShoppingItem invalidShoppingItem = new ShoppingItem
```
Comment (ShriHumrudha) [inline ShoppingList.Tests.Unit/Services/Foundations/ShoppingItems/ShoppingItemServiceTests.Validations.Add.cs]: var

**Review by ShriHumrudha — CHANGES_REQUESTED** [review]

File: ShoppingList.ConsoleApp/Models/ShoppingItems/Exceptions/FailedShoppingItemServiceException.cs (hunk 1 above)
Comment (tharaniannamalai) [reply ShoppingList.ConsoleApp/Models/ShoppingItems/Exceptions/FailedShoppingItemServiceException.cs]: Comment Resolved.

File: ShoppingList.ConsoleApp/Models/ShoppingItems/Exceptions/InvalidShoppingItemException.cs (hunk 2 above)
Comment (tharaniannamalai) [reply ShoppingList.ConsoleApp/Models/ShoppingItems/Exceptions/InvalidShoppingItemException.cs]: Comment Resolved.

File: ShoppingList.ConsoleApp/Services/Foundations/ShoppingItems/ShoppingItemService.Exception.cs (hunk 4 above)
Comment (tharaniannamalai) [reply ShoppingList.ConsoleApp/Services/Foundations/ShoppingItems/ShoppingItemService.Exception.cs:47]: Comment Resolved.

File: ShoppingList.ConsoleApp/Services/Foundations/ShoppingItems/ShoppingItemService.Exception.cs (hunk 5 above)
Comment (tharaniannamalai) [reply ShoppingList.ConsoleApp/Services/Foundations/ShoppingItems/ShoppingItemService.Exception.cs]: Comment Resolved.

File: ShoppingList.ConsoleApp/Services/Foundations/ShoppingItems/ShoppingItemService.cs (hunk 6 above)
Comment (tharaniannamalai) [reply ShoppingList.ConsoleApp/Services/Foundations/ShoppingItems/ShoppingItemService.cs]: Comment Resolved.

File: ShoppingList.Tests.Unit/Services/Foundations/ShoppingItems/ShoppingItemServiceTests.Exceptions.Add.cs (hunk 7 above)
Comment (tharaniannamalai) [reply ShoppingList.Tests.Unit/Services/Foundations/ShoppingItems/ShoppingItemServiceTests.Exceptions.Add.cs]: Comment Resolved.

File: ShoppingList.Tests.Unit/Services/Foundations/ShoppingItems/ShoppingItemServiceTests.Logic.Add.cs (hunk 8 above)
Comment (tharaniannamalai) [reply ShoppingList.Tests.Unit/Services/Foundations/ShoppingItems/ShoppingItemServiceTests.Logic.Add.cs]: Comment Resolved.

File: ShoppingList.Tests.Unit/Services/Foundations/ShoppingItems/ShoppingItemServiceTests.Validations.Add.cs (hunk 9 above)
Comment (tharaniannamalai) [reply ShoppingList.Tests.Unit/Services/Foundations/ShoppingItems/ShoppingItemServiceTests.Validations.Add.cs]: Comment Resolved.

File: ShoppingList.Tests.Unit/Services/Foundations/ShoppingItems/ShoppingItemServiceTests.Validations.Add.cs [hunk 10]
```diff
@@ -47,7 +47,7 @@ public void ShouldThrowValidationExceptionOnAddIfShoppingItemIsNullAndLogIt()
             string invalidText)
         {
             // given
-            ShoppingItem invalidShoppingItem = new ShoppingItem
+           ShoppingItem invalidShoppingItem = new ShoppingItem
```
Comment (ShriHumrudha) [inline ShoppingList.Tests.Unit/Services/Foundations/ShoppingItems/ShoppingItemServiceTests.Validations.Add.cs]: ?

File: ShoppingList.ConsoleApp/Services/Foundations/ShoppingItems/ShoppingItemService.Validations.cs [hunk 11]
```diff
@@ -0,0 +1,60 @@
...
+namespace ShoppingList.ConsoleApp.Services.Foundations.ShoppingItems
+{
+    public partial class ShoppingItemService
+    {
+        private static void ValidateShoppingItem(ShoppingItem shoppingItem)
+        {
+            ValidateShoppingItemIsNotNull(shoppingItem);
+
+            Validate(
+                (Rule: IsInvalid(shoppingItem.Id), Parameter: nameof(ShoppingItem.Id)),
```
Comment (ShriHumrudha) [inline ShoppingList.ConsoleApp/Services/Foundations/ShoppingItems/ShoppingItemService.Validations.cs]: parameter aliasing

File: ShoppingList.ConsoleApp/Services/Foundations/ShoppingItems/ShoppingItemService.cs [hunk 12]
```diff
@@ -0,0 +1,31 @@
...
+using ShoppingList.ConsoleApp.Brokers.Storages;
+using ShoppingList.ConsoleApp.Models.ShoppingItems;
+
+namespace ShoppingList.ConsoleApp.Services.Foundations.ShoppingItems
+{
+    public partial class ShoppingItemService : IShoppingItemService
+    {
+        private readonly IStorageBroker storageBroker;
+        private readonly ILoggingBroker loggingBroker;
+        public ShoppingItemService(
```
Comment (ShriHumrudha) [inline ShoppingList.ConsoleApp/Services/Foundations/ShoppingItems/ShoppingItemService.cs:16]: new line break

File: ShoppingList.Tests.Unit/Services/Foundations/ShoppingItems/ShoppingItemServiceTests.Exceptions.Add.cs [hunk 13]
```diff
@@ -0,0 +1,57 @@
...
+namespace ShoppingList.Tests.Unit.Services.Foundations.ShoppingItems
+{
+    public partial class ShoppingItemServiceTests
+    {
+        [Fact]
+        public void ShouldThrowServiceExceptionOnAddIfServiceErrorOccursAndLogIt()
+        {
+            //given
+            ShoppingItem someShoppingItem = CreateRandomShoppingItem();
+
```
Comment (ShriHumrudha) [inline ShoppingList.Tests.Unit/Services/Foundations/ShoppingItems/ShoppingItemServiceTests.Exceptions.Add.cs]: ?

File: ShoppingList.Tests.Unit/Services/Foundations/ShoppingItems/ShoppingItemServiceTests.Exceptions.Add.cs [hunk 14]
```diff
@@ -0,0 +1,57 @@
...
+            var failedShoppingItemServiceException = 
+                new FailedShoppingItemServiceException(serviceException);
+
+            var expectedShoppingItemServiceException =
+                new ShoppingItemServiceException(failedShoppingItemServiceException);
+
+            this.storageBrokerMock.Setup(broker =>
+                broker.InsertShoppingItem(It.IsAny<ShoppingItem>()))
+                    .Throws(serviceException);
+
```
Comment (ShriHumrudha) [inline ShoppingList.Tests.Unit/Services/Foundations/ShoppingItems/ShoppingItemServiceTests.Exceptions.Add.cs:35]: ?

File: ShoppingList.Tests.Unit/Services/Foundations/ShoppingItems/ShoppingItemServiceTests.Logic.Add.cs [hunk 15]
```diff
@@ -0,0 +1,44 @@
...
+using Moq;
+using ShoppingList.ConsoleApp.Models.ShoppingItems;
+using Xunit;
+
+namespace ShoppingList.Tests.Unit.Services.Foundations.ShoppingItems
+{
+    public partial class ShoppingItemServiceTests
+    {
+        [Fact]
+
```
Comment (ShriHumrudha) [inline ShoppingList.Tests.Unit/Services/Foundations/ShoppingItems/ShoppingItemServiceTests.Logic.Add.cs]: ?

File: ShoppingList.Tests.Unit/Services/Foundations/ShoppingItems/ShoppingItemServiceTests.Logic.Add.cs [hunk 16]
```diff
@@ -0,0 +1,44 @@
...
+        public void ShouldAddShoppingItem()
+        {
+            // given
+
```
Comment (ShriHumrudha) [inline ShoppingList.Tests.Unit/Services/Foundations/ShoppingItems/ShoppingItemServiceTests.Logic.Add.cs]: ?

File: ShoppingList.Tests.Unit/Services/Foundations/ShoppingItems/ShoppingItemServiceTests.Validations.Add.cs [hunk 17]
```diff
@@ -0,0 +1,91 @@
...
+            // when
+            Action addShoppingItemTask = () => this.shoppingItemService.AddShoppingItem(nullShoppingItem);
+
+            // then
+            Assert.Throws<ShoppingItemValidationException>(addShoppingItemTask);
+
+            this.loggingBrokerMock.Verify(broker =>
+               broker.LogError(It.Is(SameExceptionAs(
+                   expectedShoppingItemValidationException))),
+                    Times.Once);
```
Comment (ShriHumrudha) [inline ShoppingList.Tests.Unit/Services/Foundations/ShoppingItems/ShoppingItemServiceTests.Validations.Add.cs]: ?

File: ShoppingList.Tests.Unit/Services/Foundations/ShoppingItems/ShoppingItemServiceTests.Validations.Add.cs [hunk 18]
```diff
@@ -0,0 +1,91 @@
...
+
+            this.storageBrokerMock.Verify(broker =>
+                broker.InsertShoppingItem(It.IsAny<ShoppingItem>()),
+                Times.Never);
```
Comment (ShriHumrudha) [inline ShoppingList.Tests.Unit/Services/Foundations/ShoppingItems/ShoppingItemServiceTests.Validations.Add.cs]: ?

File: ShoppingList.Tests.Unit/Services/Foundations/ShoppingItems/ShoppingItemServiceTests.Validations.Add.cs [hunk 19]
```diff
@@ -0,0 +1,91 @@
...
+                        Times.Once);
+
+            this.storageBrokerMock.Verify(broker =>
+               broker.InsertShoppingItem(It.IsAny<ShoppingItem>()),
+                    Times.Never);
+
+            this.loggingBrokerMock.VerifyNoOtherCalls();
+            this.storageBrokerMock.VerifyNoOtherCalls();
+        }
+
```
Comment (ShriHumrudha) [inline ShoppingList.Tests.Unit/Services/Foundations/ShoppingItems/ShoppingItemServiceTests.Validations.Add.cs]: ?

File: ShoppingList.Tests.Unit/Services/Foundations/ShoppingItems/ShoppingItemServiceTests.cs [hunk 20]
```diff
@@ -0,0 +1,49 @@
...
+               && actualException.InnerException.Message == expectedException.InnerException.Message
+               && (actualException.InnerException as Xeption).DataEquals(expectedException.InnerException.Data);
+        }
+
+        private ShoppingItem CreateRandomShoppingItem() =>
+            CreateShoppingItemFiller().Create();
+
+        private static Filler<ShoppingItem> CreateShoppingItemFiller() =>
+            new Filler<ShoppingItem>();
+
```
Comment (ShriHumrudha) [inline ShoppingList.Tests.Unit/Services/Foundations/ShoppingItems/ShoppingItemServiceTests.cs]: ?

**Review by ShriHumrudha — CHANGES_REQUESTED** [review]

File: ShoppingList.Tests.Unit/Services/Foundations/ShoppingItems/ShoppingItemServiceTests.Validations.Add.cs (hunk 10 above)
Comment (tharaniannamalai) [reply ShoppingList.Tests.Unit/Services/Foundations/ShoppingItems/ShoppingItemServiceTests.Validations.Add.cs]: Comment Resolved.

Comment (ShriHumrudha) [reply ShoppingList.Tests.Unit/Services/Foundations/ShoppingItems/ShoppingItemServiceTests.Validations.Add.cs]: Please read - https://github.com/hassanhabib/CSharpCodingStandard/blob/master/Variables.md

File: ShoppingList.ConsoleApp/Services/Foundations/ShoppingItems/ShoppingItemService.cs (hunk 12 above)
Comment (tharaniannamalai) [reply ShoppingList.ConsoleApp/Services/Foundations/ShoppingItems/ShoppingItemService.cs:16]: Comment Resolved.

File: ShoppingList.Tests.Unit/Services/Foundations/ShoppingItems/ShoppingItemServiceTests.Validations.Add.cs (hunk 10 above)
Comment (tharaniannamalai) [reply ShoppingList.Tests.Unit/Services/Foundations/ShoppingItems/ShoppingItemServiceTests.Validations.Add.cs]: Comment Resolved.

File: ShoppingList.Tests.Unit/Services/Foundations/ShoppingItems/ShoppingItemServiceTests.Exceptions.Add.cs (hunk 13 above)
Comment (tharaniannamalai) [reply ShoppingList.Tests.Unit/Services/Foundations/ShoppingItems/ShoppingItemServiceTests.Exceptions.Add.cs]: Comment Resolved.

File: ShoppingList.Tests.Unit/Services/Foundations/ShoppingItems/ShoppingItemServiceTests.Exceptions.Add.cs (hunk 14 above)
Comment (tharaniannamalai) [reply ShoppingList.Tests.Unit/Services/Foundations/ShoppingItems/ShoppingItemServiceTests.Exceptions.Add.cs:35]: Comment Resolved.

File: ShoppingList.Tests.Unit/Services/Foundations/ShoppingItems/ShoppingItemServiceTests.Logic.Add.cs (hunk 15 above)
Comment (tharaniannamalai) [reply ShoppingList.Tests.Unit/Services/Foundations/ShoppingItems/ShoppingItemServiceTests.Logic.Add.cs]: Comment Resolved.

File: ShoppingList.Tests.Unit/Services/Foundations/ShoppingItems/ShoppingItemServiceTests.Logic.Add.cs (hunk 16 above)
Comment (tharaniannamalai) [reply ShoppingList.Tests.Unit/Services/Foundations/ShoppingItems/ShoppingItemServiceTests.Logic.Add.cs]: Comment Resolved.

File: ShoppingList.Tests.Unit/Services/Foundations/ShoppingItems/ShoppingItemServiceTests.Validations.Add.cs (hunk 17 above)
Comment (tharaniannamalai) [reply ShoppingList.Tests.Unit/Services/Foundations/ShoppingItems/ShoppingItemServiceTests.Validations.Add.cs]: Comment Resolved.

File: ShoppingList.Tests.Unit/Services/Foundations/ShoppingItems/ShoppingItemServiceTests.Validations.Add.cs (hunk 18 above)
Comment (tharaniannamalai) [reply ShoppingList.Tests.Unit/Services/Foundations/ShoppingItems/ShoppingItemServiceTests.Validations.Add.cs]: Comment Resolved.

File: ShoppingList.Tests.Unit/Services/Foundations/ShoppingItems/ShoppingItemServiceTests.Validations.Add.cs (hunk 19 above)
Comment (tharaniannamalai) [reply ShoppingList.Tests.Unit/Services/Foundations/ShoppingItems/ShoppingItemServiceTests.Validations.Add.cs]: Comment Resolved.

File: ShoppingList.Tests.Unit/Services/Foundations/ShoppingItems/ShoppingItemServiceTests.cs (hunk 20 above)
Comment (tharaniannamalai) [reply ShoppingList.Tests.Unit/Services/Foundations/ShoppingItems/ShoppingItemServiceTests.cs]: Comment Resolved.

File: ShoppingList.ConsoleApp/Services/Foundations/ShoppingItems/ShoppingItemService.Validations.cs (hunk 11 above)
Comment (ShriHumrudha) [reply ShoppingList.ConsoleApp/Services/Foundations/ShoppingItems/ShoppingItemService.Validations.cs]: this is still active

**Review by ShriHumrudha — CHANGES_REQUESTED** [review]

File: ShoppingList.ConsoleApp/Services/Foundations/ShoppingItems/ShoppingItemService.Validations.cs (hunk 11 above)
Comment (tharaniannamalai) [reply ShoppingList.ConsoleApp/Services/Foundations/ShoppingItems/ShoppingItemService.Validations.cs]: Comment Resolved.

File: ShoppingList.Tests.Unit/Services/Foundations/ShoppingItems/ShoppingItemServiceTests.Validations.Add.cs [hunk 21]
```diff
@@ -81,11 +81,10 @@ public void ShouldThrowValidationExceptionOnAddIfShoppingItemIsNullAndLogIt()
 
             this.storageBrokerMock.Verify(broker =>
                broker.InsertShoppingItem(It.IsAny<ShoppingItem>()),
-                    Times.Never);
+                Times.Never);
```
Comment (ShriHumrudha) [inline ShoppingList.Tests.Unit/Services/Foundations/ShoppingItems/ShoppingItemServiceTests.Validations.Add.cs]: Need a tab here

**Review by ShriHumrudha — CHANGES_REQUESTED** [review]

File: ShoppingList.Tests.Unit/Services/Foundations/ShoppingItems/ShoppingItemServiceTests.Validations.Add.cs (hunk 21 above)
Comment (tharaniannamalai) [reply ShoppingList.Tests.Unit/Services/Foundations/ShoppingItems/ShoppingItemServiceTests.Validations.Add.cs]: Comment Resolved.

**Review by ShriHumrudha — APPROVED** [review]

**Pull request merged**